operator/docs/enhancements/ruler_support.md

```diff
@@ -0,0 +1,686 @@
+...
+}
+
+// LokiTemplateSpec defines the template of all requirements to configure
+// scheduling of all Loki components to be deployed.
+type LokiTemplateSpec struct {
+...
+    // Ruler defines the ruler component spec.
+    //
+    // +optional
+    // +kubebuilder:validation:Optional
+    // +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Ruler pods"
+    Ruler *LokiComponentSpec `json:"ruler,omitempty"`
+}
+
+// LokiStackComponentStatus defines the map of per pod status per LokiStack component.
+// Each component is represented by a separate map of v1.Phase to a list of pods.
+type LokiStackComponentStatus struct {
+...
+    // Ruler is a map to the per pod status of the lokistack ruler statefulset.
+    //
+    // +optional
+    // +kubebuilder:validation:Optional
+    // +operator-sdk:csv:customresourcedefinitions:type=status,xDescriptors="urn:alm:descriptor:com.tectonic.ui:podStatuses",displayName="Ruler",order=6
+    Ruler PodStatusMap `json:"ruler,omitempty"`
+}
+```
+
+#### LokiRule definition
+
+The `LokiRule` CRD comprises a set of specifications and status definitions to declare groups of alerting and/or recording rules for a single `LokiStack` instance. The syntax for the rule groups resembles the official [Prometheus Rule syntax](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules). In addition the status definition provides support for rule validation conditions:
+
+1. If a `LokiRule` includes an alert name and a record name it is ambiguous if it is an alerting or a recording rule.
+2. If a `LokiRule` includes an invalid `for` period it is an invalid alerting rule.
+3. If a `LokiRule` includes an invalid `record` metric name it is an invalid recording rule.
+4. If a `LokiRule` includes an invalid LogQL `expr` it is an invalid rule.
+5. If a `LokiRule` includes two groups with the same name it is an invalid rule.
+6. If none of above applies a `LokiRule` is considered a valid rule.
+
+```go
+// EvaluationDuration defines the type for Prometheus durations.
+//
+// +kubebuilder:validation:Pattern:="((([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?|0)"
+type EvaluationDuration string
+
+// LokiRuleSpec defines the desired state of LokiRule
+type LokiRuleSpec struct {
+    // Name of the LokiStack to reconcile the rules for
+    //
+    // +required
+    // +kubebuilder:validation:Required
+    StackName string `json:"stackName"`
+
+    // List of groups for alerting and/or recording rules.
+    //
+    // +optional
+    // +kubebuilder:validation:Optional
+    Groups []*LokiRuleGroup `json:"groups"`
+}
+
+// LokiRuleGroup defines a group of Loki alerting and/or recording rules.
+type LokiRuleGroup struct {
+    // Name defines a name of the present recoding/alerting rule. Must be unique
+    // within all loki rules.
+    //
+    // +required
+    // +kubebuilder:validation:Required
+    Name string `json:"name"`
```

Comment:
> > Must be unique within all loki rules
> 
>     * What are the consequences of duplicate group names? (warning, duplicates are ignored, hard error)
> 

I am not sure where/why Prometheus maintains this limitation and Loki adopts it. But I consider that rule group names are used like IDs from tools like cortextool, e.g. examplary output from cortextool:
```
Namespace                                                                              | Rule Group
openshift-operators-redhat-loki-operator-dev-163b29f5-bc65-4289-9457-c8f94269d0e2.yaml | should_operator_record
openshift-operators-redhat-loki-operator-dev-d908b623-b32e-4cd6-8800-75323199c706.yaml | should_operator_fire_on_errors
openshift-operators-redhat-loki-operator-dev-d908b623-b32e-4cd6-8800-75323199c706.yaml | credentials_operator_leak
```
Currently the proposed validation webhook would disallow creating multiple groups in the same rule.

>     * Would it be an idea to auto-generate the group names in the operator based on the namespace/name of the `LokiRule`? What about only having one group per `LokiRule`?

I suggest to support user-defined group names as well as multiple groups inside a rule to maintain similiarity as close as possible to `PrometheusRule`.

Reply:
> Currently the proposed validation webhook would disallow creating multiple groups in the same rule.

Yes, that should be possible. Also groups with the same name across multiple rules, but this makes the validation harder and less transparent. The offending other rule might be in a different namespace that I as a user do not have write access to but want to use in my Loki using the label-based configuration. This might also mean that the validation hook would need to know about that configuration (which LokiStack matches which Rules, so that conflicts can be found correctly).

> I suggest to support user-defined group names as well as multiple groups inside a rule to maintain similiarity as close as possible to PrometheusRule.

Similarity to the existing prometheus-operator / PrometheusRule is a valid argument. I need to check if and how the issue is addressed there.

For me, auto-generated group names still sound like a convenient and efficient way of getting rid of the validation issue.

Reply:
> > Currently the proposed validation webhook would disallow creating multiple groups in the same rule.
> 
> Yes, that should be possible. Also groups with the same name across multiple rules, but this makes the validation harder and less transparent. The offending other rule might be in a different namespace that I as a user do not have write access to but want to use in my Loki using the label-based configuration. This might also mean that the validation hook would need to know about that configuration (which LokiStack matches which Rules, so that conflicts can be found correctly).
> 

The requirement of unique group names applies only within **a single** rule and not across rules. TBH as far as I understand tooling handling/output, e.g. `cortextool` the rule name is creating some sort of namespace and within this all group names are required to be unique, e.g.:

```
Namespace                                                                              | Rule Group
openshift-operators-redhat-loki-operator-dev-d908b623-b32e-4cd6-8800-75323199c706.yaml | should_operator_fire_on_errors
openshift-operators-redhat-loki-operator-dev-d908b623-b32e-4cd6-8800-75323199c706.yaml | credentials_operator_leak
```

> > I suggest to support user-defined group names as well as multiple groups inside a rule to maintain similiarity as close as possible to PrometheusRule.
> 
> Similarity to the existing prometheus-operator / PrometheusRule is a valid argument. I need to check if and how the issue is addressed there.
> 
> For me, auto-generated group names still sound like a convenient and efficient way of getting rid of the validation issue.

I believe auto-generating group names is defeating the purpose of users giving names that mean something to them and their workloads. As longs as we don't need to provide uniqueness across rules, I believe we can leave this field for user edits.